### PR TITLE
Fix data race in discovery filter

### PR DESCRIPTION
**Pull request by howardjohn**
Example crash https://prow.istio.io/view/gs/istio-prow/logs/integ-k8s-126_istio_postsubmit/1790453700397895680

### CODE REVIEW

File: pkg/kube/namespace/filter.go [hunk 1]
```diff
@@ -79,8 +79,13 @@ func NewDiscoveryNamespacesFilter(
 			}
 		},
 		DeleteFunc: func(ns *corev1.Namespace) {
-			// No need to notify handlers for deletes
-			f.namespaceDeleted(ns.ObjectMeta)
+			f.lock.Lock()
+			defer f.lock.Unlock()
```
Comment (hzxuzhonghu) [inline pkg/kube/namespace/filter.go:83]: LGTM, I think this is the critical fix.

But what is the target for updating  pkg/kube/kclient/client.go

File: pkg/kube/namespace/filter.go [hunk 2]
```diff
@@ -194,10 +199,9 @@ func (d *discoveryNamespacesFilter) namespaceUpdatedLocked(oldNs, newNs metav1.O
 	return false, false
 }
 
-// namespaceDeleted : if deleted namespace was a member, remove it
-func (d *discoveryNamespacesFilter) namespaceDeleted(ns metav1.ObjectMeta) (membershipChanged bool) {
+// namespaceDeletedLocked : if deleted namespace was a member, remove it
+func (d *discoveryNamespacesFilter) namespaceDeletedLocked(ns metav1.ObjectMeta) {
```
Comment (hzxuzhonghu) [inline pkg/kube/namespace/filter.go:203]: nit: we donot need this function or move the lock here

File: pkg/kube/kclient/client.go [hunk 3]
```diff
@@ -333,17 +333,16 @@ func applyDynamicFilter[T controllers.ComparableObject](filter Filter, gvr schem
...
-							c.handler.OnAdd(item, false)
-						} else {
-							c.handler.OnDelete(item)
-						}
+						c.handler.OnAdd(item, false)
 					}
 				}
+				// Removes are currently NOT handled. We only have the namespace name here. We would need to have the object
+				// filter passthrough the entire namespace object, so we can pass the last known state to OnDelete.
+				// Fortunately, missing a namespace delete event usually doesn't matter since everything in the namespace gets torn down.
```
Comment (hzxuzhonghu) [inline pkg/kube/kclient/client.go:345]: Does this mean there is no namespace delete triggered? I think this is not right, this package can be used by not only istiod, but also other thirdparty users

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]

File: pkg/kube/kclient/client.go (hunk 3 above)
Comment (howardjohn) [reply pkg/kube/kclient/client.go:345]: That is the state even before this PR. Istio packages are not intended for usage outside istio so I don't really care much about the 3p use case? 

We can make it send deletes consistently but it's extremely hard to do so. The problem is the filter and the namespace handler ordering is random (from informer) so we cannot guarantee the order. I don't think the complexity is worth it imo. 

Either way IMO we could merge this even if we want to do this eventually, as it is strictly fixing the bug -- the missing deletes was already present

**Review by hzxuzhonghu — APPROVED** [review]
Agreed, but better to improve. I have know many projects import istio libs

**Pull request merged**